### PR TITLE
add operator built-in webhook

### PR DESCRIPTION
**Pull request by calvin0327**
Signed-off-by: calvin0327 <wen.chen@daocloud.io>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add operator built-in webhook.

**Which issue(s) this PR fixes**:
Part of # https://github.com/karmada-io/karmada/issues/2979

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: operator/cmd/operator/app/operator.go [hunk 1]
```diff
@@ -90,6 +91,12 @@ func Run(ctx context.Context, o *options.Options) error {
 		return err
 	}
 
+	klog.Info("registering webhook to the manager")
+	if err := (&operatorv1alpha1.Karmada{}).SetupWebhookWithManager(manager); err != nil {
```
Comment (jwcesign) [inline operator/cmd/operator/app/operator.go:95]: It's a little wired to create a new empty struct to call the function.

File: operator/cmd/operator/app/options/options.go [hunk 2]
```diff
@@ -28,6 +28,13 @@ type Options struct {
 	// SecurePort is the port that the the server serves at.
 	// Note: We hope support https in the future once controller-runtime provides the functionality.
 	SecurePort int
+	// CertDir is the directory that contains the server key and certificate.
+	// if not set, webhook server would look up the server key and certificate in {TempDir}/k8s-webhook-server/serving-certs.
+	CertDir string
+	// CertName is the server certificate name. Defaults to tls.crt.
+	CertName string
```
Comment (jwcesign) [inline operator/cmd/operator/app/options/options.go:35]: how about CertFile/KeyFile? more explicit

File: operator/pkg/apis/operator/v1alpha1/karmada_webhook.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+	"k8s.io/apimachinery/pkg/api/validation"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// SetupWebhookWithManager new a webhook for karmada resource and register to manager.
```
Comment (jwcesign) [inline operator/pkg/apis/operator/v1alpha1/karmada_webhook.go:17]: SetupWebhookWithManager creates a new webhook for karmada resource and register to manager

File: operator/pkg/apis/operator/v1alpha1/karmada_webhook_test.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+									ImageTag:        "v1.4.0",
+								},
+								Replicas: pointer.Int32(1),
+							},
+						},
+						KarmadaWebhook: &KarmadaWebhook{
+							CommonSettings: CommonSettings{
+								Image: Image{
+									ImageRepository: "127.0.0.1:8080/karmada/karmada-webhook",
+									ImageTag:        "v1.4.0",
```
Comment (jwcesign) [inline operator/pkg/apis/operator/v1alpha1/karmada_webhook_test.go:188]: will it be affected by the changed version? like a new release?

**Comment by RainbowMango** [comment]
`Fixes ` will close the referencing issue. Please update it to `Part of` if it is partial work.

**Comment by calvin0327** [comment]
I will close the pr, please see https://github.com/karmada-io/karmada/pull/2992

**Comment by calvin0327** [comment]
Thanks for review.

**Comment by calvin0327** [comment]
It is not worth developing webhooks just to fill in the initial values. I will close the pr.

**Pull request closed without merging**